src/core/pubsubhub.js

```diff
@@ -74,10 +77,40 @@ export function unsub({ topic, cb }) {
 
 sub("error", err => {
   console.error(err, err.stack);
+  if (onRespecError) onRespecError(err);
 });
 
 sub("warn", str => {
   console.warn(str);
+  if (onRespecWarn) onRespecWarn(str);
 });
 
 expose(name, { sub });
+
+export async function run(conf) {
+  if (conf.collectErrors && !conf.hasOwnProperty("onRespecError")) {
+    // Automatically assign a callback function to collect errors.
+    conf.onRespecError = err => {
+      if (!conf.hasOwnProperty("collectedErrors"))
+        conf.collectedErrors = new Set();
+      const { collectedErrors } = conf;
+      collectedErrors.add(err);
+    };
+  }
+  if (conf.collectWarnings && !conf.hasOwnProperty("onRespecWarn")) {
```

Comment:
Thanks to Swapnil's observation, I came up with this slightly dirty trick to collect Errors and Warnings. 
We can't pass the promises as we planned earlier. I think this is similar to why we need [pre-process-spec file](https://github.com/w3c/respec/blob/develop/tests/spec/core/pre-process-spec.html)

Reply:
#886 needs something similar. It should be a good way to integrate that as well.